[Bug]: [NIXL] Hetero TP assertion fails when tp > num_kv_heads (GQA replication)

### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

text
vLLM version: 0.21.0
Python version: 3.12.11
OS: Linux 6.12.58-82.121.amzn2023.x86_64 x86_64


</details>


### Describe the bug

_validate_remote_agent_handshake in NixlConnectorWorker (worker.py:~1498) asserts:

python
assert remote_block_len == (self.block_len_per_layer[0] * tp_ratio) // block_size_ratio


This assumes block_len scales linearly with 1/tp_size. However, when tp_size > num_kv_heads, GQA replication caps num_kv_heads_per_rank at 1, so block_len stops  scaling. Both sides end up with the same block_len despite different TP sizes, causing the assertion to fail.

### Steps to reproduce

Run disaggregated inference with:
- Model: Any model with few KV heads (e.g., meta-llama/Llama-3.2-1B-Instruct — 8 KV heads)
- Prefill: --tensor-parallel-size 8
- Decode: --tensor-parallel-size 16
- KV connector: NixlConnector

### Error

AssertionError: Remote P worker KV layer cache must be of shape [2, N, local_kv_heads*tp_ratio, page_size, head_dim] and same dtype.


### Configurations tested (model with 8 KV heads)

| Config | D_TP > kv_heads? | P_TP > kv_heads? | Result |
|--------|---|---|---|
| tp2→tp4 | No | No | PASS |
| tp4→tp8 | No (edge: 8=8) | No | PASS |
| tp8→tp16 | Yes (16>8) | No (edge) | FAIL |
| tp4→tp16 | Yes (16>8) | No |  FAIL |
| tp16→tp8 | No (edge) | Yes (16>8) | FAIL |

### Root cause

When tp > num_kv_heads, num_kv_heads_per_rank = max(1, num_kv_heads // tp) = 1 (GQA replication). Block_len is proportional to num_kv_heads_per_rank, so it caps at
a single head's worth. The assertion expects remote_block_len = local_block_len × tp_ratio but both sides report the same block_len due to this cap.

### Suggested fix

The validation should account for GQA replication. For example, skip the strict proportionality check when max(local_tp, remote_tp) > total_num_kv_heads:

```
if tp_ratio > 0:
    expected = (self.block_len_per_layer[0] * tp_ratio) // block_size_ratio
    if remote_block_len != expected:
        # Allow when GQA replication caps block_len on either side
        total_kv_heads = self.transfer_topo.total_num_kv_heads
        if max(self.transfer_topo.tp_size, remote_tp_size) > total_kv_heads:
            pass  # block_len doesn't scale beyond kv_head count
        else:
            assert False, (
                "Remote P worker KV layer cache must be of shape [2, N,"
                " local_kv_heads*tp_ratio, page_size, head_dim] and same dtype."
            )

```



### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

E2E verification complete - GQA replication works correctly with the fix applied.

Setup:
- Hardware: 8× A100-SXM4-40GB
- Model: TinyLlama/TinyLlama-1.1B-Chat-v1.0 (4 KV heads, 32 attn heads)
- Config: Prefill TP=2 → Decode TP=8 (tp8 > 4 kv_heads triggers GQA replication)
- vLLM 0.21.0 with remote_heads // local_heads fix applied to _validate_remote_agent_handshake
- Disaggregated via NIXL + toy_proxy_server.py

Without fix:
`AssertionError: Remote P worker KV layer cache must be of shape [2, N, local_kv_heads*tp_ratio, page_size, head_dim] and same dtype.
`

With fix, e2e request through disaggregated pipeline:
```
curl http://localhost:8300/v1/completions -d '{"model":"TinyLlama/TinyLlama-1.1B-Chat-v1.0","prompt":"The capital of France is","max_tokens":10}'

{"choices":[{"text":" Paris Business language in Stock market has a significant role","finish_reason":"length"}]}

```
Handshake passes, KV transfers via NIXL, decode generates coherent output starting with "Paris". Confirms data correctness -  corrupted KV under replication would produce garbage.

@waynehacking8 your test wrapper script works as designed on 8×A100. Happy to run additional configs if needed.


Validated against the same fix approach used in #45879 (@Oseltamivir) and #45860 (mine) - all three PRs use remote_heads // local_heads instead of raw tp_ratio. LGTM on #45879 which has comprehensive testing. Happy to see it land.

--- a/tests/v1/kv_connector/unit/test_nixl_connector.py
+++ b/tests/v1/kv_connector/unit/test_nixl_connector.py
@@ -517,17 +517,17 @@ class FakeNixlConnectorWorker(NixlConnectorWorker):
         assert expected_engine_id == self.REMOTE_ENGINE_ID
 
         # Adjust remote block length metadata to satisfy heterogeneous TP
-        # invariants enforced during handshake validation.
+        # invariants enforced during handshake validation.  Use per-rank
+        # head ratio (not tp_ratio) to account for GQA replication capping.
         remote_block_lens = list(self.block_len_per_layer)
         tp_ratio = self.transfer_topo.tp_ratio(remote_tp_size)
-        if remote_tp_size > self.world_size:
-            # P TP > D TP case, block_len of remote is smaller
+        total_kv = self.transfer_topo.total_num_kv_heads
+        local_heads = self.transfer_topo.local_physical_heads
+        remote_heads = max(1, total_kv // remote_tp_size)
+        if remote_tp_size != self.world_size:
             remote_block_lens = [
-                block_len // (-tp_ratio) for block_len in remote_block_lens
-            ]
-        elif remote_tp_size < self.world_size:
-            remote_block_lens = [
-                block_len * tp_ratio for block_len in remote_block_lens
+                block_len * remote_heads // local_heads
+                for block_len in remote_block_lens
             ]
 
         # When remote tp_size > local tp_size, handshake with multiple
@@ -1143,6 +1143,123 @@ class TestNixlHandshake:
             )
             with pytest.raises(AssertionError):
                 worker2.add_remote_agent(bad_meta, remote_tp_size=1)
+
+    @patch(
+        "vllm.distributed.kv_transfer.kv_connector.v1.nixl.base_worker.NixlWrapper",
+        FakeNixlWrapper,
+    )
+    def test_handshake_validates_gqa_replicated_block_len(
+        self, default_vllm_config, dist_init
+    ):
+        """Regression test for #45330.
+
+        When tp_size > total_num_kv_heads, GQA replication caps per-rank
+        KV heads at 1, so block_len stops scaling with 1/tp.  With 8 KV
+        heads and D_TP=16 pulling from P_TP=8, both sides hold one head
+        per rank and report the *same* block_len; the old validation
+        expected local_block_len * tp_ratio and rejected the valid
+        handshake.
+        """
+        vllm_config = create_vllm_config()
+
+        with patch(
+            "vllm.distributed.kv_transfer.kv_connector.v1.nixl.base_worker.get_tensor_model_parallel_world_size",  # noqa: E501
+            return_value=16,
+        ):
+            connector = NixlConnector(
+                vllm_config,
+                KVConnectorRole.WORKER,
+                make_kv_cache_config(block_size=16),
+            )
+            connector.connector_worker = FakeNixlConnectorWorker(
+                vllm_config, connector.engine_id, hand_shake_latency=0
+            )
+            worker = connector.connector_worker
+
+            worker.transfer_topo.total_num_kv_heads = 8
+            worker.transfer_topo.local_physical_heads = 1
+            worker.kv_cache_layout = "HND"
+
+            worker.slot_size_per_layer = [4096]
+            worker.block_len_per_layer = [4096 * worker.block_size]
+            worker.num_blocks = 1
+            worker.dst_num_blocks[worker.engine_id] = worker.num_blocks
+
+            # Remote P with TP=8 also has 1 head/rank -> identical
+            # block_len despite tp_ratio == 2.
+            meta = NixlAgentMetadata(
+                engine_id=FakeNixlConnectorWorker.REMOTE_ENGINE_ID,
+                agent_metadata=FakeNixlWrapper.AGENT_METADATA,
+                kv_caches_base_addr=[0],
+                device_id=0,
+                num_blocks=1,
+                block_lens=list(worker.block_len_per_layer),
+                kv_cache_layout="HND",
+                block_size=worker.block_size,
+                ssm_sizes=(0, 0),
+                attn_backend_name=worker.backend_name,
+                physical_blocks_per_logical_kv_block=1,
+            )
+
+            # Must validate cleanly (used to raise AssertionError).
+            worker.add_remote_agent(meta, remote_tp_size=8)
+
+    @patch(
+        "vllm.distributed.kv_transfer.kv_connector.v1.nixl.base_worker.NixlWrapper",
+        FakeNixlWrapper,
+    )
+    def test_handshake_rejects_wrong_block_len_without_gqa_replication(
+        self, default_vllm_config, dist_init
+    ):
+        """Ensure the head-ratio validation still rejects genuinely wrong
+        block_lens when GQA replication is NOT in effect (32 KV heads,
+        D_TP=4, P_TP=2: head_ratio=4, both sides have >1 head/rank).
+        """
+        vllm_config = create_vllm_config()
+
+        with patch(
+            "vllm.distributed.kv_transfer.kv_connector.v1.nixl.base_worker.get_tensor_model_parallel_world_size",  # noqa: E501
+            return_value=4,
+        ):
+            connector = NixlConnector(
+                vllm_config,
+                KVConnectorRole.WORKER,
+                make_kv_cache_config(block_size=16),
+            )
+            connector.connector_worker = FakeNixlConnectorWorker(
+                vllm_config, connector.engine_id, hand_shake_latency=0
+            )
+            worker = connector.connector_worker
+
+            worker.transfer_topo.total_num_kv_heads = 32
+            worker.transfer_topo.local_physical_heads = 8  # 32 // 4
+            worker.kv_cache_layout = "HND"
+
+            slot_size = 4096
+            worker.slot_size_per_layer = [slot_size]
+            worker.block_len_per_layer = [slot_size * worker.block_size]
+            worker.num_blocks = 1
+            worker.dst_num_blocks[worker.engine_id] = worker.num_blocks
+
+            # Remote P_TP=2 has 16 heads/rank -> head_ratio = 16/8 = 2.
+            # Correct remote block_len = local * 2.  Send local * 1
+            # (wrong) to verify rejection.
+            bad_meta = NixlAgentMetadata(
+                engine_id=FakeNixlConnectorWorker.REMOTE_ENGINE_ID,
+                agent_metadata=FakeNixlWrapper.AGENT_METADATA,
+                kv_caches_base_addr=[0],
+                device_id=0,
+                num_blocks=1,
+                block_lens=list(worker.block_len_per_layer),
+                kv_cache_layout="HND",
+                block_size=worker.block_size,
+                ssm_sizes=(0, 0),
+                attn_backend_name=worker.backend_name,
+                physical_blocks_per_logical_kv_block=1,
+            )
+
+            with pytest.raises(AssertionError):
+                worker.add_remote_agent(bad_meta, remote_tp_size=2)
 
 
 # NOTE: resource cleanup in mp backend is a bit finicky, so the order in which

--- a/vllm/distributed/kv_transfer/kv_connector/v1/nixl/base_worker.py
+++ b/vllm/distributed/kv_transfer/kv_connector/v1/nixl/base_worker.py
@@ -1596,7 +1596,9 @@ class NixlBaseConnectorWorker:
         # Per-region block_len validation enforcing the P/D invariant.
         # REPLICATE regions (MLA, or a whole-model MLA / replicated-KV transfer)
         # only allow the number of blocks to differ; SPLIT regions scale with
-        # tp_ratio. Mamba uses the ssm_sizes counterpart, so skip block_len here.
+        # the per-rank KV head ratio rather than the raw tp_ratio, because GQA
+        # replication caps per-rank heads at 1 when tp > total_kv_heads
+        # (issue #45330). Mamba uses the ssm_sizes counterpart, so skip here.
         if not self._has_mamba:
             assert len(self.block_len_per_layer) == len(nixl_agent_meta.block_lens), (
                 "Number of KV layers must match between prefill and decode"
@@ -1604,6 +1606,9 @@ class NixlBaseConnectorWorker:
             model_replicated = self.use_mla or self.transfer_topo.is_kv_replicated(
                 remote_engine_id
             )
+            total_kv_heads = self.transfer_topo.total_num_kv_heads
+            local_heads = self.transfer_topo.local_physical_heads
+            remote_heads = max(1, total_kv_heads // remote_tp_size)
             for i, local_len in enumerate(self.block_len_per_layer):
                 replicated = model_replicated or self._is_region_replicated(i)
                 remote_len = nixl_agent_meta.block_lens[i]
@@ -1614,9 +1619,13 @@ class NixlBaseConnectorWorker:
                         f"remote={remote_len}, bsr={block_size_ratio})."
                     )
                 elif tp_ratio > 0:
-                    assert remote_len == (local_len * tp_ratio) // block_size_ratio, (
+                    assert (
+                        remote_len
+                        == (local_len * remote_heads // local_heads) // block_size_ratio
+                    ), (
                         f"SPLIT region {i}: remote P KV block_len {remote_len} "
-                        f"must equal local {local_len} * tp_ratio {tp_ratio} "
+                        f"must equal local {local_len} * remote_heads "
+                        f"{remote_heads} // local_heads {local_heads} "
                         f"// block_size_ratio {block_size_ratio}."
                     )
                 else:
@@ -1624,10 +1633,10 @@ class NixlBaseConnectorWorker:
                         "Different local/remote block sizes are not supported "
                         "when P TP > D TP."
                     )
-                    assert remote_len == local_len // (-tp_ratio), (
-                        f"SPLIT region {i}: remote P KV block_len "
-                        f"{remote_len} must equal local {local_len} "
-                        f"// |tp_ratio| {-tp_ratio}."
+                    assert remote_len == local_len * remote_heads // local_heads, (
+                        f"SPLIT region {i}: remote P KV block_len {remote_len} "
+                        f"must equal local {local_len} * remote_heads "
+                        f"{remote_heads} // local_heads {local_heads}."
                     )
 
         # TP workers that handhshake with same remote have same #blocks.

